packages/plugin-npm-cli/sources/commands/npm/auditUtils.ts

```diff
@@ -0,0 +1,69 @@
+import {DescriptorHash, LocatorHash, Project, Workspace} from '@yarnpkg/core';
+
+// Enumerate all the transitive dependencies of a set of top-level packages
+function getTransitiveDependencies(project: Project, roots: Array<DescriptorHash>): Set<DescriptorHash> {
+  // Queue of dependency descriptorHashes to visit; set of already-visited patterns
+  const queue: Array<DescriptorHash> = [];
+  const descriptorHashes = new Set<DescriptorHash>();
+
+  const enqueue = (descriptorHash: DescriptorHash) => {
+    if (descriptorHashes.has(descriptorHash))
+      return;
+
+    descriptorHashes.add(descriptorHash);
+    queue.push(descriptorHash);
+  };
+
+  roots.forEach(enqueue);
+
+  // Final result set
+  const transitiveDependencies = new Set<DescriptorHash>();
+
+  const {storedResolutions, originalPackages} = project;
+
+  while (queue.length > 0) {
+    const descriptorHash = queue.shift();
+    const locatorHash = storedResolutions.get(descriptorHash as DescriptorHash);
```

Comment:
I'm having a bit of trouble here. For some reason none of the descriptor hashes passed in from `roots` have had entries in `storedResolutions`. It keeps returning `undefined` so there is no `originalPackage` found, and the rest of the code in this loop is skipped.

Reply:
Usually this is caused when you forget to run `await project.restoreInstallState()` at the beginning of your Command implementation (cf `run.ts` for an example).

Reply:
Ohh I certainly didn't do this. Thanks for the tip!

Reply:
I got a bit closer on this issue today.

I found that when I try to use a `descriptorHash` from `workspace.manifest.dependencies.values()` to get a `locatorHash` from `project.storedResolutions`, it always returns `undefined`.
But if I use `workspace.dependencies.values()` instead, it is able to retrieve the locator hashes.

`workspace.dependencies` includes dependencies + devDependencies + workspaces. And for this part I need to be able to distinguish dependencies from devDependencies, so I can't really use it without some other way of identifying the dependency type.

If anyone has any ideas why this is happening, let me know. I'll keep plugging away in the meantime.

Reply:
Descriptors from `manifest.dependencies` are the raw descriptors, exactly as listed in the `package.json` file. Because we normalize them before resolving them (`foo@1.2.3` -> `foo@npm:1.2.3`), they probably aren't part of the `storedResolution` map.

> And for this part I need to be able to distinguish dependencies from devDependencies, so I can't really use it without some other way of identifying the dependency type.

You probably can do a bit of both: use the `workspace.dependencies` set, then check whether the dependencies' ident is part of `workspace.manifest.devDependencies`.

Reply:
Ahh right. I didn't realise there was a difference as they used the same types.
Thanks for the pointer! This is working now.